src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -772,6 +773,37 @@ MergeTreeDataSelectSamplingData MergeTreeDataSelectExecutor::getSampling(
     return sampling;
 }
 
+std::optional<std::unordered_set<String>> MergeTreeDataSelectExecutor::filterPartsByVirtualColumns(
+    const MergeTreeData & data,
+    const MergeTreeData::DataPartsVector & parts,
+    const ActionsDAGPtr & filter_dag,
+    ContextPtr context)
+{
+    if (!filter_dag)
+        return {};
+
+    auto sample = data.getSampleBlockWithVirtualColumns();
+    std::unordered_set<const ActionsDAG::Node *> allowed_inputs;
+    for (const auto * input : filter_dag->getInputs())
+        if (sample.has(input->result_name))
+            allowed_inputs.insert(input);
+
+    if (allowed_inputs.empty())
+        return {};
+
+    auto atoms = filter_dag->extractConjunctionAtoms(filter_dag->getOutputs().at(0));
+    atoms = ActionsDAG::filterNodesByAllowedInputs(std::move(atoms), allowed_inputs);
```

Comment:
@KochetovNicolai this will not work if the query will have something except for columns from partition by, right?

```sql
CREATE TABLE default.forms
(
    `form_id` FixedString(24),
    `text_field` String
)
ENGINE = MergeTree
PRIMARY KEY form_id
ORDER BY form_id
SETTINGS index_granularity = 8192

insert into forms values ('5840ead423829c1eab29fa97','this is a test');
```

```sql
p620.local :) explain indexes=1 select * from forms where _part like '%foo%' and text_field like '%foo%' settings allow_experimental_analyzer=1

EXPLAIN indexes = 1
SELECT *
FROM forms
WHERE (_part LIKE '%foo%') AND (text_field LIKE '%foo%')
SETTINGS allow_experimental_analyzer = 1

Query id: e23f014a-546e-4eab-9fe4-5c09e8ba2c13

┌─explain───────────────────────────────────┐
│ Expression ((Project names + Projection)) │
│   Filter                                  │
│     ReadFromMergeTree (default.forms)     │
│     Indexes:                              │
│       PrimaryKey                          │
│         Condition: true                   │
│         Parts: 1/1                        │
│         Granules: 1/1                     │
└───────────────────────────────────────────┘
```

```sql
p620.local :) explain indexes=1 select * from forms where _part like '%foo%' settings allow_experimental_analyzer=1

EXPLAIN indexes = 1
SELECT *
FROM forms
WHERE _part LIKE '%foo%'
SETTINGS allow_experimental_analyzer = 1

Query id: 053ce7a4-fbdb-4f26-86f1-8e3adf8d0c26

┌─explain────────────────────────────────────────────────────────┐
│ Expression ((Project names + Projection))                      │
│   Filter ((WHERE + Change column names to column identifiers)) │
│     ReadFromMergeTree (default.forms)                          │
└────────────────────────────────────────────────────────────────┘

3 rows in set. Elapsed: 0.002 sec.

```